### PR TITLE
[Feature]: speedup autoscheme with multi-process and cache files

### PR DESCRIPTION
**Pull request by xin3he**
### Feature Description
Save and use cached per-op score files to skip time-consuming sensitivity evaluation.
Enable multi-process to speedup the evaluation. It will fallback if multi-process fails, like OOM.
`export AR_ENABLE_AUTO_SCHEME_PARALLEL=1`

### Motivation and Use Case
For LLM, like GLM5.2, the AutoScheme usually takes 1h.
For different bits and shared_layers, we usually need to re-run the entire AutoScheme process, which is not necessary if we cache the per-op score.

### Real model test
1. auto-round /models/Qwen3.6-35B-A3B/ --options "MXFP4,MXFP8"  --target_bits 4.3
2. auto-round /models/Qwen3.6-35B-A3B/ --options "MXFP4,MXFP8"  --target_bits 4.5  --shared_layers q_proj k_proj v_proj 
3. auto-round /models/Qwen3-32B/ --options "MXFP4,MXFP8"  --target_bits 4.5 
4. auto-round /models/Qwen3-32B/ --options "MXFP4,MXFP8"  --target_bits 4.6  --shared_layers q_proj k_proj v_proj 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/xinhe/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/xinhe/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

IDX | Main 耗时 | This Branch 耗时 | 耗时变化 | Main RAM | This Branch RAM | RAM 变化 | Main VRAM | This Branch VRAM | VRAM 变化
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 29m 53s | 18m 30s | ↓ 11m 23s（38.1%） | 70.82 GB | 140.45 GB | ↑ 69.63 GB（98.3%） | 10.23 GB | 20.45 GB | ↑ 10.22 GB（99.9%）
2 | 23m 27s | 38s | ↓ 22m 49s（97.3%） | 70.93 GB | 4.31 GB | ↓ 66.62 GB（93.9%） | 10.23 GB | 2.76 GB | ↓ 7.47 GB（73.0%）
3 | 8m 01s | 6m 49s | ↓ 1m 12s（15.0%） | 67.23 GB | 132.68 GB | ↑ 65.45 GB（97.4%） | 14.36 GB | 28.97 GB | ↑ 14.61 GB（101.7%）
4 | 7m 02s | 1s | ↓ 7m 01s（99.8%） | 67.27 GB | 6.92 GB | ↓ 60.35 GB（89.7%） | 14.36 GB | 2.90 GB | ↓ 11.46 GB（79.8%）

</body>

</html>

## Type of Change

New feature

## Related Issues

Fixes or relates to #2080 #1988

## Checklist Before Submitting

- [ ] My code has been tested locally.
- [ ] Documentation has been updated as needed.
- [ ] New or updated tests are included where applicable.
- [ ] The CUDA CI has passed. You can trigger it by commenting `/azp run Unit-Test-CUDA-AutoRound`.

### CODE REVIEW

File: auto_round/auto_scheme/delta_loss.py [hunk 1]
```diff
@@ -1903,19 +2473,22 @@ def _select_embedding_scheme_index():
...
-        min_dp_avg = min_dp_bits / max(sum(layer_numel.get(n, 0) for n in total_scores), 1)
-        raise ValueError(
-            f"target_bits is infeasible: after subtracting fixed/embedding layers the remaining "
-            f"budget ({target_params_cnt} bits) is below the minimum the given options can reach "
-            f"({min_dp_bits} bits, avg {min_dp_avg:.3f}). Raise target_bits, add a lower-bit option, "
-            f"or lower the bits of fixed layers."
-        )
+    logger.info(
+        "AutoScheme post-scoring: DP selection took %.2fs (layers=%d)",
+        time.perf_counter() - dp_started,
```
[the remark this answers is not in the document]
Comment (xin3he) [reply auto_round/auto_scheme/delta_loss.py:2612]: Accepted. Added an explicit infeasible-budget check immediately after `choose_bits_per_layer_with_path`. AutoScheme now raises the intended `ValueError("Avg bits is too small")` before attempting to iterate a missing path.

File: auto_round/auto_scheme/delta_loss.py [hunk 2]
```diff
@@ -1432,6 +1464,340 @@ def _apply_head_trick(head_name, schemes, sorted_indices, target_bits, target_pa
...
+        {
+          "version": 3,
+          "score_granularity": "per_op",
+          "cache_key": "<hex>",
+          "scheme_index": 0,
+          "scheme": { ... scheme dict ... },
+          "created_at": "<ISO datetime>",
+          "layer_scores": { layer_key: [[bits, loss], ...], ... },
+          "total_loss_for_scheme": 1.234,
+          "total_params": 12345
```
[the remark this answers is not in the document]
Comment (xin3he) [reply auto_round/auto_scheme/delta_loss.py:1563]: Accepted. Updated the version 3 cache schema docstring so each `layer_scores` entry is documented as the single `[bits, loss]` pair that the serializer and tests actually use.

File: auto_round/auto_scheme/delta_loss.py [hunk 3]
```diff
@@ -1670,108 +2036,312 @@ def check_bf16_scheme(scheme):
...
+                total_params=sum(
+                    (
+                        module.weight.numel()
+                        if hasattr(module, "weight") and module.weight is not None
+                        else getattr(module, "_cached_weight_numel", 0)
+                    )
+                    for name, module in model.named_modules()
+                    if name in quant_layer_names + embedding_layers_names
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (xin3he) [reply auto_round/auto_scheme/delta_loss.py:2175]: Accepted. Precomputed `scored_layer_names` once outside `_save_per_op_scores` and reused the set while iterating `model.named_modules()`, reducing each membership check from linear to constant time.

File: auto_round/utils/device.py [hunk 4]
```diff
@@ -1026,66 +1026,36 @@ def partition_dict_numbers(number_dict, n):
...
-        for _ in range(n - len(number_dict)):
-            groups.append({})
-        return groups
-
-    if n == len(number_dict):
-        return [{key: value} for key, value in number_dict.items()]
-
-    total_sum = sum(number_dict.values())
-    # target = total_sum / n  # Use float for better precision
+    # Quick handling of edge cases
```
[the remark this answers is not in the document]
Comment (xin3he) [reply auto_round/utils/device.py:1028]: Accepted. Removed the stale `itertools.combinations` import; the rewritten greedy partition implementation no longer references it.

File: auto_round/utils/model.py [hunk 5]
```diff
@@ -925,6 +925,112 @@ def model_save_pretrained(model, save_directory, **kwargs):
...
+                device=device,
+                model_dtype=model_dtype,
+                trust_remote_code=trust_remote_code,
+            )
+            if use_auto_mapping is not None:
+                _kwargs["use_auto_mapping"] = use_auto_mapping
+            _kwargs.update(kwargs)
+            model, processor, tokenizer, image_processor = mllm_load_model(pretrained_model_name_or_path, **_kwargs)
+            model = _prepare_model(model)
+            return model, tokenizer, processor, image_processor, None, True, False
```
Comment (n1ck-guo) [inline auto_round/utils/model.py]: I suggest declaring initial values—such as `model = None` and `_is_mllm = False`—and returning the parameter names when returning values （return model, _is_mllm instead of return model,m True, False）this would improve readability. It returns many values, making it difficult to remember what each one represents.

File: auto_round/utils/model.py [hunk 6]
```diff
@@ -925,6 +925,112 @@ def model_save_pretrained(model, save_directory, **kwargs):
     return pipe, model.to(device)
 
 
+def load_model(
```
Comment (n1ck-guo) [inline auto_round/utils/model.py:939]: Since a unified interface already exists, I think it would be better to replace `mllm_load` and `llm_load` with `load_model` and rename to `_mllm_load`. However, if you feel this involves too significant a change, I can also handle it in a separate PR later.

File: auto_round/utils/device_manager.py [hunk 7]
```diff
@@ -1055,7 +1070,15 @@ def get_device_memory(i: int = 0) -> int:
...
-    return dev_mgr.total_memory(i) / 1024 / 1024 / 1024
+
+    total_memory = dev_mgr.total_memory(i)
+    if total_memory is None:
+        raise RuntimeError(f"Failed to query total memory for device index {i} on backend {dev_mgr.type}.")
+    if total_memory <= 0:
+        raise RuntimeError(
+            f"Detected non-positive total memory ({total_memory}) for device index {i} on backend {dev_mgr.type}."
+        )
+    return total_memory / 1024 / 1024 / 1024
```
Comment (n1ck-guo) [inline auto_round/utils/device_manager.py]: total_memory / 1024**3?

File: auto_round/auto_scheme/delta_loss.py [hunk 8]
```diff
@@ -1432,6 +1464,371 @@ def _apply_head_trick(head_name, schemes, sorted_indices, target_bits, target_pa
...
+        scheme=scheme,
+    )
+    refreshed_scores = {}
+    for name, (_, loss) in cached_layer_scores.items():
+        bits, _ = compute_layer_bits(get_module(model, name), ignore_scale_zp_bits)
+        refreshed_scores[name] = [bits, loss]
+    return refreshed_scores
+
+
+def _assign_scheme_worker_devices(worker_count, visible_gpu_count):
```
Comment (n1ck-guo) [inline auto_round/auto_scheme/delta_loss.py]: Personally, I donot like this kind of function (it contains very little code and is only called once). I feel it would be better to just write the logic directly into the main workflow. However, that’s a matter of coding style—it’s up to you.

File: auto_round/utils/model.py (hunk 5 above)
Comment (xin3he) [reply auto_round/utils/model.py]: Accepted. Initialized every return field and model-type flag with a descriptive name, changed each dispatch branch to assign those variables, and consolidated the function to one named 7-tuple return. This preserves the existing API while making the return order explicit.

File: auto_round/utils/model.py (hunk 6 above)
Comment (xin3he) [reply auto_round/utils/model.py:939]: Thank you for the suggestion. I am keeping `llm_load_model` and `mllm_load_model` as the specialized loaders: the unified `load_model` dispatches to them, so replacing their implementations with `load_model` would recurse. They are also directly used by `ModelContext` and existing public imports/tests, so renaming them to private functions would be a broader API migration unrelated to the AutoScheme worker integration in this PR.

File: auto_round/auto_scheme/delta_loss.py (hunk 8 above)
Comment (xin3he) [reply auto_round/auto_scheme/delta_loss.py]: Thank you for the style suggestion. I am keeping `_assign_scheme_worker_devices` separate because it owns the logical-device round-robin rule and its invalid-GPU-count validation, while the caller coordinates pool setup. The helper also has three focused tests covering multiple GPUs, one GPU, and the zero-GPU error; inlining it would couple those boundary tests to the much larger multiprocessing workflow without changing behavior.

**Comment by n1ck-guo** [comment]
LGTM

**Review by n1ck-guo — APPROVED** [review]
